Fix/refactor decode_base64_field processor (libbeat)

guyromb opened the pull request:
Fixes:
- Key is not fetched correctly
- Clean duplicated doc key in processors-using
- Aligning AllowedFields according to spec
- General cleaning/refactoring

#docs #review #needs_backport #libbeat #bug

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

guyromb commented:
run jenkins

guyromb commented:
jenkins, test this please

guyromb commented:
Could anyone from elastic be kind to review && hotfix this? This feature is broken.
@dedemorton @ph

kvch commented on libbeat/processors/actions/decode_base64_field.go:
```diff
@@ -46,34 +45,28 @@ type base64Config struct {
 	processors.RegisterPlugin(processorName,
 		checks.ConfigChecked(NewDecodeBase64Field,
 			checks.RequireFields("field"),
-			checks.AllowedFields("field", "when")))
```
Please don't delete `when`. It is required for conditional processing.

guyromb replied:
fixed

kvch commented on libbeat/processors/actions/decode_base64_field.go:
```diff
@@ -84,55 +77,55 @@ func (f *decodeBase64Field) Run(event *beat.Event) (*beat.Event, error) {
+		return fmt.Errorf("could not fetch base64 value for key: %s, Error: %v", f.config.Field.From, err)
 	}
 
-	text, ok := value.(string)
```
Please do not rename `ok`, as it's a Golang convention when doing type assertions.

guyromb replied:
fixed

kvch commented on libbeat/docs/processors-using.asciidoc:
```diff
@@ -215,7 +215,6 @@ ifdef::has_decode_csv_fields_processor[]
  * <<decode-csv-fields,`decode_csv_fields`>>
 endif::[]
  * <<decode-json-fields,`decode_json_fields`>>
- * <<decode-base64-field,`decode_base64_field`>>
```
Why did you remove this reference?

guyromb replied:
It is duplicated :-) I just removed the second reference.

kvch commented on libbeat/processors/actions/decode_base64_field.go:
```diff
@@ -84,55 +77,55 @@ func (f *decodeBase64Field) Run(event *beat.Event) (*beat.Event, error) {
-	if to == "" || from == to {
+	if f.config.Field.To == "" || f.config.Field.From == f.config.Field.To {
 		// Deletion must happen first to support cases where a becomes a.b
-		if err = fields.Delete(from); err != nil {
```
Do you mind removing this `if`? It's not required in this processor, but got copied from rename processor.

guyromb replied:
done

guyromb commented:
@kvch I addressed the issues. 
Do you know if to support base64 processor using hints (k8s auto discovery) we also need to do any additional changes?

kvch commented:
jenkins test this

kvch commented:
Tested locally, seems to work,

kvch commented:
It seems that the error messages in the unit test need to be adjusted.

guyromb commented:
On it @kvch

kvch commented:
jenkins test this

guyromb commented:
Seems like it fails due to unrelated things (connection timed out) both on travis and jenkins.

kvch commented:
@guyromb Thank you for your contribution!
